Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/WANDPowderReduction.py

```diff
@@ -88,72 +148,206 @@ def PyExec(self):
-        ConvertSpectrumAxis(InputWorkspace='__data_tmp', Target=target, EFixed=eFixed, OutputWorkspace=outWS, EnableLogging=False)
-        Transpose(InputWorkspace=outWS, OutputWorkspace=outWS, EnableLogging=False)
-        ResampleX(InputWorkspace=outWS, OutputWorkspace=outWS, XMin=xMin, XMax=xMax, NumberBins=numberBins, EnableLogging=False)
-
-        if cal is not None:
-            ExtractUnmaskedSpectra(InputWorkspace=cal, MaskWorkspace='__mask_tmp', OutputWorkspace='__cal_tmp', EnableLogging=False)
-            if isinstance(mtd['__cal_tmp'], IEventWorkspace):
-                Integration(InputWorkspace='__cal_tmp', OutputWorkspace='__cal_tmp', EnableLogging=False)
-            CopyInstrumentParameters(data, '__cal_tmp', EnableLogging=False)
-            ConvertSpectrumAxis(InputWorkspace='__cal_tmp', Target=target, EFixed=eFixed, OutputWorkspace='__cal_tmp', EnableLogging=False)
-            Transpose(InputWorkspace='__cal_tmp', OutputWorkspace='__cal_tmp', EnableLogging=False)
-            ResampleX(InputWorkspace='__cal_tmp', OutputWorkspace='__cal_tmp', XMin=xMin, XMax=xMax, NumberBins=numberBins,
-                      EnableLogging=False)
-            Divide(LHSWorkspace=outWS, RHSWorkspace='__cal_tmp', OutputWorkspace=outWS, EnableLogging=False)
-            if normaliseBy == "Monitor":
-                cal_scale = cal.run().getProtonCharge()
-            elif normaliseBy == "Time":
-                cal_scale = cal.run().getLogData('duration').value
-
-        Scale(InputWorkspace=outWS, OutputWorkspace=outWS, Factor=cal_scale/data_scale, EnableLogging=False)
-
-        if bkg is not None:
-            ExtractUnmaskedSpectra(InputWorkspace=bkg, MaskWorkspace='__mask_tmp', OutputWorkspace='__bkg_tmp', EnableLogging=False)
-            if isinstance(mtd['__bkg_tmp'], IEventWorkspace):
-                Integration(InputWorkspace='__bkg_tmp', OutputWorkspace='__bkg_tmp', EnableLogging=False)
-            CopyInstrumentParameters(data, '__bkg_tmp', EnableLogging=False)
-            ConvertSpectrumAxis(InputWorkspace='__bkg_tmp', Target=target, EFixed=eFixed, OutputWorkspace='__bkg_tmp', EnableLogging=False)
-            Transpose(InputWorkspace='__bkg_tmp', OutputWorkspace='__bkg_tmp', EnableLogging=False)
-            ResampleX(InputWorkspace='__bkg_tmp', OutputWorkspace='__bkg_tmp', XMin=xMin, XMax=xMax, NumberBins=numberBins,
-                      EnableLogging=False)
+        _get_scale = lambda x: {
+            None: 1,
+            "Monitor": x.run().getProtonCharge(),
+            "Time": x.run().getLogData("duration").value,
+        }[normaliseBy]
+
+        # NOTE:
+        # StringArrayProperty cannot be optional, so the background can only be passed in as a string
+        # or a list, which will be manually unpacked here
+        try:
+            bkg = bkg.split(",") if bkg is not None else bkg
+        except:
+            _bg = AnalysisDataService.retrieve(
```

Comment:
What if bkg[0] is not a workspace in AnalysisDataService?

Reply:
My original plan was to use ` StringArrayProperty` for background since it should have 1-1 mapping with the input spectrum (InputWorkspace).
But it seems like I cannot make ` StringArrayProperty` an optional input as it lack corresponding support.
The backup plan here is to force users to use string or list of strings to indicate the backgrounds.
If `AnalysisDataService` fails to retrieve the workspace, the back-trace error should be sufficient for users to realize that correct backgrounds are needed in order to move forward.

Reply:
After digging around a little, I have to settle with using `str` for background input as I cannot find an input option from Mantid that will allow heterogeneous input.
The current interface allow users to use either `Workspace2D`, `list` or `str` for InputWorkspace (`StringArrayProperty`, required), and `str` for Background.